p2p/src/peer_manager/tests/connections.rs

```diff
@@ -434,3 +436,122 @@ async fn inbound_connection_too_many_peers_mock() {
     )
     .await;
 }
+
+async fn connection_timeout<T>(addr1: T::Address, addr2: T::Address)
+where
+    T: NetworkingService + 'static + std::fmt::Debug,
+    T::ConnectivityHandle: ConnectivityService<T>,
+    <T as net::NetworkingService>::Address: std::str::FromStr,
+    <<T as net::NetworkingService>::Address as std::str::FromStr>::Err: std::fmt::Debug,
+{
+    let config = Arc::new(config::create_mainnet());
+    let mut swarm1 = make_peer_manager::<T>(addr1, Arc::clone(&config)).await;
+
+    swarm1
+        .peer_connectivity_handle
+        .connect(addr2.clone())
+        .await
+        .expect("dial to succeed");
+
+    match timeout(
+        Duration::from_secs(swarm1._p2p_config.outbound_connection_timeout),
+        swarm1.peer_connectivity_handle.poll_next(),
+    )
+    .await
+    {
+        Ok(res) => assert!(std::matches!(
+            res,
+            Ok(net::types::ConnectivityEvent::ConnectionError {
+                address: _,
+                error: P2pError::DialError(DialError::IoError(
+                    std::io::ErrorKind::ConnectionRefused
```

Comment:
Why is a timed out connection showing as a refused connection?

Reply:
Getting the correct error out would require either making all of the errors generic over the network service or having a leaky abstraction where the global error types would contain a `Multiaddr`. Admittedly there is some nuance lost if we treat connection refused and connection timeout both as connection refused but the result is handled the same way in both cases: we retry later again or we abandon this peer completely. Is it ok if I log the dial error using `log:.error!()` and keep this at least for now as `ConnectionRefused`?

Reply:
Yes. Please log it correctly and change the error label to "ConnectionRefusedOrTimedout"